### PR TITLE
[PY] chore: refactor _on_turn

### PR DESCRIPTION
**Pull request by BMS-geodev**
## Linked issues

closes: #1559

## Details

Broke up the _on_turn() method for clarity. The logical parts of the large functionality were broken into separate neighboring private methods each of which having more atomic responsibilities. 

The behavior of _on_turn() should remain consistent, but in general be more readable. 

#### Change details

N/A

**code snippets**:
Here are some sample adjustments. 

ex. remove mentions:
```
async def _on_turn(self, context: TurnContext):
...
    # remove @mentions
    if self.options.remove_recipient_mention:
        if context.activity.type == ActivityTypes.message:
            context.activity.text = context.remove_recipient_mention(context.activity)
```
is now:
```
async def _on_turn(self, context: TurnContext):
...
    await self._remove_mentions(context)
```
```
async def _remove_mentions(self, context: TurnContext):
    if self.options.remove_recipient_mention and context.activity.type == ActivityTypes.message:
        context.activity.text = context.remove_recipient_mention(context.activity.text)
```

ex. running the AI chain when matches are not found:
```
async def _on_turn(self, context: TurnContext):
...
    # only run chain when no activity handlers matched
    if (
        matches == 0
        and self._ai
        and self._options.ai
        and context.activity.type == ActivityTypes.message
        and context.activity.text
    ):
        is_ok = await self._ai.run(context, state)
    
        if not is_ok:
            await state.save(context, self._options.storage)
            return
```
is now:
```
async def _on_turn(self, context: TurnContext):
    if matches == 0:
      if not await self._run_ai_chain(context, state):
          return
```
```
async def _run_ai_chain(self, context: TurnContext, state):
    if (
        self._ai
        and self._options.ai
        and context.activity.type == ActivityTypes.message
        and context.activity.text
    ):
        is_ok = await self._ai.run(context, state)
        if not is_ok:
            await self._save_state(context, state)
            return False
    return True
```

**screenshots**:

## Attestation Checklist

- [X] My code follows the style guidelines of this project

- I have checked for/fixed spelling, linting, and other errors
- I have commented my code for clarity
- I have made corresponding changes to the documentation (updating the doc strings in the code is sufficient)
- My changes generate no new warnings
- I have added tests that validates my changes, and provides sufficient test coverage. I have tested with:
  - Local testing
  - E2E testing in Teams
- New and existing unit tests pass locally with my changes

### Additional information

> Feel free to add other relevant information below

### CODE REVIEW

**Comment by BMS-geodev** [comment]
Naming things is tough, especially when new to a codebase. Please let me know if any naming choices clash with any style guidelines or intended vocabulary.

**Comment by BMS-geodev** [comment]
@microsoft-github-policy-service agree company="Awarity Inc."

**Comment by BMS-geodev** [comment]
Local testing, as well as `poetry run test` both positive.

**Comment by corinagum** [comment]
Do tests need to be updated?

**Comment by BMS-geodev** [comment]
@corinagum , I pulled the most recent changes from main and locally ran `poetry run test` and got all successful. I believe the PR is in a good state, so much as I can see.

File: python/packages/ai/teams/app.py [hunk 1]
```diff
@@ -647,76 +647,93 @@ async def on_turn(self, context: TurnContext):
...
+
+    async def _run_after_turn_middleware(self, context: TurnContext, state):
+        for after_turn in self._after_turn:
+            is_ok = await after_turn(context, state)
+            if not is_ok:
+                await self._save_state(context, state)
+                return False
+        return True
+
+    async def _save_state(self, context: TurnContext, state):
```
Comment (aacebo) [inline python/packages/ai/teams/app.py:734]: im not sure abstracting this single line function call to another method that calls it adds much value

Comment (BMS-geodev) [reply python/packages/ai/teams/app.py:734]: sure, happy to roll that back in. As I mentioned at the top, I went totally verbose on the breakout, not yet knowing the goals or new tests that might be anticipated. Will change shortly.

Comment (BMS-geodev) [reply python/packages/ai/teams/app.py]: rolled state.save back and tested locally.

File: python/packages/ai/teams/app.py [hunk 2]
```diff
@@ -664,115 +664,90 @@ async def on_turn(self, context: TurnContext):
...
-                input_files = state.temp.input_files if state.temp.input_files is not None else []
-                for file_downloader in self._options.file_downloaders:
-                    files = await file_downloader.download_files(context)
-                    input_files.append(files)
-                state.temp.input_files = input_files
-
-            # run activity handlers
-            is_ok, matches = await self._on_activity(context, state)
+            await self._start_typing(context)
+            await self._remove_mentions(context)
```
Comment (lilyydu) [inline python/packages/ai/teams/app.py:668]: can remove async/await - should be synchronous

File: python/packages/ai/teams/app.py [hunk 3]
```diff
@@ -664,115 +664,90 @@ async def on_turn(self, context: TurnContext):
...
+            state = await self._initialize_state(context)
+            if not await self._run_before_turn_middleware(context, state):
+                return
 
+            await self._handle_file_downloads(context, state)
+            is_ok, matches = await self._run_activity_handlers(context, state)
```
Comment (lilyydu) [inline python/packages/ai/teams/app.py:674]: we can likely keep this section of running activity handlers the same - function extracted is a bit redundant

File: python/packages/ai/teams/app.py [hunk 4]
```diff
@@ -664,115 +664,90 @@ async def on_turn(self, context: TurnContext):
...
-
-            state: StateT = cast(StateT, await TurnState.load(context, self._options.storage))
-
-            if self._turn_state_factory:
-                state = await self._turn_state_factory(context)
-
-            await state.load(context, self._options.storage)
-            state.temp.input = context.activity.text
-
-            # authentication
```
Comment (lilyydu) [inline python/packages/ai/teams/app.py:683]: believe you may have missed this auth chunk!

**Comment by BMS-geodev** [comment]
still chasing updates on other branches, ensuring code here is synced

**Comment by BMS-geodev** [comment]
After merging in the state, and auth changes from other PR's, and making the needed adjustments, I now have working poetry tests, as well as samples behaving as expected. 

I believe this PR is ready for review again.

File: python/packages/ai/teams/app.py [hunk 5]
```diff
@@ -664,115 +664,116 @@ async def on_turn(self, context: TurnContext):
...
+            state = await self._turn_state_factory(context)
+
+        await state.load(context, self._options.storage)
+        state.temp.input = context.activity.text
+        return state
+
+    async def _authenticate_user(self, context: TurnContext, state):
+        if self.options.auth and self._auth:
+            auth_condition = (
+                isinstance(self.options.auth.auto, bool) and self.options.auth.auto
```
Comment (lilyydu) [inline python/packages/ai/teams/app.py:719]: do we need to specifically check if `self.options.auth.auto` is True? same for `self.options.auth.auto(context)`

File: python/packages/ai/teams/app.py [hunk 6]
```diff
@@ -664,115 +664,116 @@ async def on_turn(self, context: TurnContext):
...
+            ) or (callable(self.options.auth.auto) and self.options.auth.auto(context))
+            user_in_sign_in = IN_SIGN_IN_KEY in state.user
+            print(auth_condition, user_in_sign_in)
```
Comment (lilyydu) [inline python/packages/ai/teams/app.py:722]: extra print statement

File: python/packages/ai/teams/app.py [hunk 7]
```diff
@@ -664,115 +664,116 @@ async def on_turn(self, context: TurnContext):
...
+            if auth_condition or user_in_sign_in:
+                key = state.user.get(IN_SIGN_IN_KEY, self.options.auth.default)
```
Comment (lilyydu) [inline python/packages/ai/teams/app.py:724]: this is the `else` case, but `if` case should set it to `state.user[IN_SIGN_IN_KEY]`

Comment (lilyydu) [reply python/packages/ai/teams/app.py:724]: also missing the check for `if key is not None`

File: python/packages/ai/teams/app.py [hunk 8]
```diff
@@ -664,115 +664,116 @@ async def on_turn(self, context: TurnContext):
...
+                res = await self._auth.sign_in(context, state, key=key)
+                print(res)
```
Comment (lilyydu) [inline python/packages/ai/teams/app.py:726]: extra print

File: python/packages/ai/teams/app.py [hunk 9]
```diff
@@ -664,114 +664,116 @@ async def on_turn(self, context: TurnContext):
...
+        if self._turn_state_factory:
+            state = await self._turn_state_factory(context)
+
+        await state.load(context, self._options.storage)
+        state.temp.input = context.activity.text
+        return state
+
+    async def _authenticate_user(self, context: TurnContext, state):
+        if self.options.auth and self._auth:
+            auth_condition = (isinstance(self.options.auth.auto, bool)) or (
```
Comment (lilyydu) [inline python/packages/ai/teams/app.py:718]: need to also check

`and self.options.auth.auto is True`
`and self.options.auth.auto(context) is True`

**Review by lilyydu — APPROVED** [review]
Thank you for the great work, edits and patience 🙌🏻 Let me know once we've tested the sample flows and we'll be good to merge!

**Pull request merged**